File: Contest/Europe/CERC/2014/D.java
import java.io.*;
import java.util.*;
import java.lang.*;
import java.math.*;

public class D {
    static Scanner in = new Scanner(System.in);
    static int n;
    static int[][] g;
    static boolean[] v;
    static BigInteger[][] ans;

    private static boolean connect(int i, int j) {
        long dx = g[i][0] - g[j][0];
        long dy = g[i][1] - g[j][1];
        long r = g[i][2] + g[j][2];
        return dx * dx + dy * dy == r * r;
    }

    private static void dfs(int now, BigInteger nu, BigInteger nd) {
        ans[now][0] = nu; ans[now][1] = nd;
        v[now] = true;

        for (int i = 0; i < n; ++i) if (!v[i] && connect(now, i)) 
            dfs(i, nu.multiply(BigInteger.valueOf(-1)).multiply(BigInteger.valueOf(g[now][2])), nd.multiply(BigInteger.valueOf(g[i][2])));
    }
    
    private static void solve() {
        n = in.nextInt();
        g = new int[n][3];
        ans = new BigInteger[n][2];
        v = new boolean[n];

        for (int i = 0; i < n; ++i) {
            g[i][0] = in.nextInt();
            g[i][1] = in.nextInt();
            g[i][2] = in.nextInt();
            ans[i][0] = ans[i][1] = BigInteger.ZERO;
            v[i] = false;
        }

        dfs(0, BigInteger.ONE, BigInteger.ONE);

        for (int i = 0; i < n; ++i) {
            if (ans[i][0] == BigInteger.ZERO) {
                System.out.println("not moving");
                continue;
            }

            BigInteger z = ans[i][0].gcd(ans[i][1]);
            ans[i][0] = ans[i][0].divide(z);
            ans[i][1] = ans[i][1].divide(z);
            
            if (ans[i][1].compareTo(BigInteger.ONE) == 1) 
                System.out.print(ans[i][0].abs() + "/" + ans[i][1].abs() + " ");
            else 
                System.out.print(ans[i][0].abs() + " ");

            System.out.println(ans[i][0].multiply(ans[i][1]).compareTo(BigInteger.ZERO) == 1 ? "clockwise" : "counterclockwise");
        }
    }
    public static void main(String[] args) {
        int t = in.nextInt();
        while (t > 0) {
            solve();
            --t;
        }
    }
}

File: Contest/Europe/CERC/2014/D.py

from math import gcd

n, g = 0, []
ans, v = [], []

def connect(i, j):
    global g
    dx = g[i][0] - g[j][0]
    dy = g[i][1] - g[j][1]
    r = g[i][2] + g[j][2]
    # print(i, j, dx, dy, r)
    return dx ** 2 + dy ** 2 == r ** 2

def dfs(now, nu, nd):
    global n, g, ans, v
    # print(now, nu, nd)
    ans[now] = [nu, nd]
    
    v[now] = True
    for i in range(n): 
        if not v[i]:
            if connect(now, i):
                dfs(i, int(-nu * g[now][2]), int(nd * g[i][2]))

def solve():
    global n, g, ans, v
    n = int(input())
    g = [list(map(int, input().split(' '))) for _ in range(n)]
    ans = [[0, 0] for _ in range(n)]
    v = [False for _ in range(n)]

    dfs(0, 1, 1)

    for i in range(n):
        if ans[i][0] == 0: 
            print("not moving")
            continue

        z = gcd(ans[i][0], ans[i][1])
        ans[i][0] /= z
        ans[i][1] /= z
        if ans[i][1] > 1: print(str(int(abs(ans[i][0]))) + "/" + str(int(abs(ans[i][1]))), end=' ')
        else: print(str(int(abs(ans[i][0]))), end=' ')
        print("clockwise" if ans[i][0] * ans[i][1] > 0 else "counterclockwise")
            
    

t = int(input())

for _ in range(t): solve()
